### PR TITLE
Remove race condition with run-test

### LINKED ISSUE #73099: The test "Simple pod should support inline execution and attach" is inherently racy.
The test case "executing a command with run and attach without stdin" is inherently racy.

The test runs`kubectl run run-test --image=docker.io/library/busybox:1.29 --restart=OnFailure --attach=true --stdin -- sh -c cat && echo 'stdin closed'`, and expected to see `stdin closed` in the output.

The case when the test passes:
* case 1:
  * Kubernetes creates and starts the container without stdin;
  * The container exits before attach runs;
  * kubectl falls back to get container log, which contains `stdin closed`. https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/run/run.go#L527
* case 2:
  * Kubernetes creates and starts the container without stdin;
  * kubectl attach before "stdin closed" is printed;
  * kubectl attach contains `stdin closed`.

The case when the test fails:
  * Kubernetes creates and starts the container without stdin;
  * The container prints "stdin closed", but hasn't exited yet;
  * kubectl attach contains nothing.

One example failure: https://gubernator.k8s.io/build/kubernetes-jenkins/logs/ci-containerd-e2e-ubuntu-gce/7057

Related issue: https://github.com/kubernetes/kubernetes/issues/24070

Can we modify this test not to check the output? Or use `kubectl log` to check output instead?

/cc @kubernetes/sig-node-bugs @kubernetes/sig-cli-bugs

### PR DESCRIPTION
**Pull request by selfmanagingresource**
**What type of PR is this?**
> /kind flake

**What this PR does / why we need it**:
This change modifies kubectl run e2e test to remove a race condition where kubectl run might complete before we can scrape the logs off via attach.
**Which issue(s) this PR fixes**:
Fixes #73099

**Special notes for your reviewer**:
Tested via running the e2e tests a few times on gce via kubetest.

**Does this PR introduce a user-facing change?**:
No
```release-note
None
```

### CODE REVIEW

**Review by aixeshunter — APPROVED** [review]

File: test/e2e/kubectl/kubectl.go [hunk 1]
```diff
@@ -514,8 +514,28 @@ var _ = SIGDescribe("Kubectl client", func() {
...
+				}
+				logOutput := framework.RunKubectlOrDie(nsFlag, "logs", runTestPod.Name)
+				Expect(runOutput).To(ContainSubstring("abcd1234"))
+				Expect(runOutput).To(ContainSubstring("stdin closed"))
+				return strings.Contains(logOutput, "abcd1234"), nil
+			})
+			if err != nil {
+				os.Exit(1)
+			}
+			Expect(err).To(BeNil())
```
Comment (oomichi) [inline test/e2e/kubectl/kubectl.go]: line 534-536 seems unnecessary because of line 537.

File: test/e2e/kubectl/kubectl.go [hunk 2]
```diff
@@ -514,8 +514,28 @@ var _ = SIGDescribe("Kubectl client", func() {
...
 				WithStdinData("abcd1234").
 				ExecOrDie()
-			Expect(runOutput).To(ContainSubstring("abcd1234"))
-			Expect(runOutput).To(ContainSubstring("stdin closed"))
+
+			g := func(pods []*v1.Pod) sort.Interface { return sort.Reverse(controller.ActivePods(pods)) }
+			runTestPod, _, err := polymorphichelpers.GetFirstPod(f.ClientSet.CoreV1(), ns, "run=run-test", 1*time.Minute, g)
+			if err != nil {
+				os.Exit(1)
+			}
```
Comment (oomichi) [inline test/e2e/kubectl/kubectl.go]: Why don't we use
`Expect(err).To(BeNil())`
to know the actual err message?

File: test/e2e/kubectl/kubectl.go (hunk 1 above)
Comment (selfmanagingresource) [reply test/e2e/kubectl/kubectl.go]: Agreed and fixed. I had copied some boilerplate from the check the at the end of the test which was already using kubectl logs for similar reasons. I went ahead and cleaned up those redundant checks as well.

File: test/e2e/kubectl/kubectl.go (hunk 2 above)
Comment (selfmanagingresource) [reply test/e2e/kubectl/kubectl.go]: Fixed

File: test/e2e/kubectl/kubectl.go (hunk 1 above)
Comment (oomichi) [reply test/e2e/kubectl/kubectl.go]: Nice catch, thanks for updating

**Review by oomichi — COMMENTED** [review]
/lgtm

**Comment by nikhita** [comment]
/kind bug

/assign @soltysh 
for approval

**Review by soltysh — APPROVED** [review]
/approve
Although, I think we should work on making this test simpler rather more complicated. @selfmanagingresource do you think you could work on something simpler?

**Comment by selfmanagingresource** [comment]
Sure I'll take a look. Do you have any code pointers where a similar refactoring was done? Specifically, were you thinking of moving more of this into integration tests and scoping down the e2e tests or just refactoring common pieces into common methods etc?

**Pull request merged**